src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs

```diff
@@ -0,0 +1,1504 @@
+                Array.Copy(bucket,target,bucket.Length)
+                target, map.getBucket() , ref (!countRef)
+                
+        let trgtBitDepth = calcBitMaskDepth ((target.Length * ShardSize) - 1)
+        let bucketBitMask = calcSubBitMask trgtBitDepth
+        
+        let mapFn = 
+            if additions.Length = target.Length then
+                // Split into two types of mapping, one-to-one, and small map expansion
+                ////////////////////////////////////////////////////////////////////////
+                fun bi si sm ->
+                    let tshrd = target.[bi]
+                    let tm = tshrd.[si]
+                    // Buckets same size so simple to one-to-one map
+                    if isNull tm then
+                        Interlocked.Add(tCount,MapTree.size sm) |> ignore
+                        tshrd.[si] <- sm
+                    else
+                        tshrd.[si] <-
+                            MapTree.fold (fun acc k v -> // for each key in source
+                                MapTree.addAndIncr comparer k v tCount acc // lookup doubled so might be worth extending Maptree to return both tree & `newAddition' bool
+                            ) tm sm
+            else
+                fun _ si sm ->
+                    // Bucket lower order so shard needs to be remapped to higher order
+                    ///////////////////////////////////////////////////////////////////////
+                    MapTree.iter (fun k v -> // for each key in source
+                        let kh = k.GetHashCode()
+                        let tbi = bucketIndex(kh,bucketBitMask)
+                        let tshrd = target.[tbi]
+                        let tm = tshrd.[si]
+
+                        if isNull tm then
+                            Interlocked.Increment(tCount) |> ignore
+                            tshrd.[si] <- MapOne(k,v)
+                        else                                             
+                            tshrd.[si] <- MapTree.addAndIncr comparer k v tCount tm
+                    ) sm
+
+
+        eachMapInBucket(additions , mapFn)
+        // Step through maps to union
+        ShardMap<'K,'V>(!tCount,target)
+
+
+    /////////////////////////////////
+    /// Static Methods
+    /////////////////////////////////
+
+    static member inline Empty with get () = ShardMap<'K,'V>(0)
+
+    static member Collect (collectFn: 'T -> #seq<'V>) (keyFn:'V -> 'K) (collection:seq<'T>) =
+        
+        let initSize = Seq.length collection //:HACK
+        let size = if initSize < ShardSize then ShardSize else initSize
+        let comparer = LanguagePrimitives.FastGenericComparer<'K>
+        let bitdepth = calcBitMaskDepth size
+        let bucketSize = bucketSizeFromBitDepth (bitdepth)
+        let bucketBitMask = calcSubBitMask bitdepth
+        let newBucket = Array.zeroCreate<Shard<'K,'V>>(bucketSize)
+        let countRef = ref 0
```

Comment:
I believe this should be ``let mutable countRef`` and you should pass the address of ``countRef`` where appropriate?  That would save an allocation

Reply:
Thank you, noted, I will update in source

Reply:
![image](https://user-images.githubusercontent.com/5388991/34045619-99a827c0-e1a1-11e7-9ef7-c617dbcdd100.png)
updating means I am then unable to use `Interlocked.Increment` on the value so not sure if I can pass a ref to update if I use `let mutable countRef` ?